### PR TITLE
Add a new optional command to export textured mesh to a USD format

### PR DESCRIPTION
**Pull request by serguei-k**
## Description

Add a new **optional** command to export textured mesh to a USD format.

## Features list

- [X] Add new command exportUSD to export mesh and textures to USD
- [X] Add a simple Materials class to capture the small subset of material properties and associated textures that is supported by texturing pipeline
- [X] Update texturing pipeline to make use of new material class
- [X] Fix bug in normal computation (checking for NaN)
- [X] Fix bug in loading mesh data not setting the correct number of atlases and materials used
- [X] Fix bug in loading mesh data using incorrect material ID due to default material
- [X] Add option to merge coincident vertices when loading meshes with multiple materials

### CODE REVIEW

File: src/aliceVision/mesh/Mesh.cpp [hunk 1]
```diff
@@ -1333,7 +1334,7 @@ void Mesh::computeNormalsForPts(StaticVector<StaticVector<int>>& ptsNeighTris, S
             {
                 Point3d n1 = computeTriangleNormal(triTmp[j]);
                 n1 = n1.normalize();
-                if(!std::isnan(n1.x) && !std::isnan(n1.y) && std::isnan(n1.z)) // check if is not NaN
+                if(!std::isnan(n1.x) && !std::isnan(n1.y) && !std::isnan(n1.z)) // check if is not NaN
```
Comment (serguei-k) [inline src/aliceVision/mesh/Mesh.cpp:1337]: Was getting 0 normals, found this small issue

File: src/aliceVision/mesh/Mesh.cpp [hunk 2]
```diff
@@ -2529,9 +2535,82 @@ void Mesh::load(const std::string& filepath)
...
+        {
+            f.v[0] = oldToNewMap[f.v[0]];
+            f.v[1] = oldToNewMap[f.v[1]];
+            f.v[2] = oldToNewMap[f.v[2]];
+        }
+    }
+
+    // set number of materials used
+    const std::unordered_set<int> materialIds = std::unordered_set<int>(_trisMtlIds.begin(), _trisMtlIds.end());
+    nmtls = static_cast<int>(materialIds.size());
```
Comment (serguei-k) [inline src/aliceVision/mesh/Mesh.cpp:2571]: nmtls property was never set, so when loading an asset with materials the number of atlases wasn't being set correctly.

File: src/aliceVision/mesh/Texturing.cpp [hunk 3]
```diff
@@ -940,6 +944,26 @@ void Texturing::loadWithAtlas(const std::string& filepath, bool flipNormals)
     }
 }
 
+void Texturing::loadWithMaterial(const std::string& filepath, bool flipNormals)
```
Comment (serguei-k) [inline src/aliceVision/mesh/Texturing.cpp:947]: Added as new method for clarity, could edit existing load method instead.

File: src/software/pipeline/main_texturing.cpp [hunk 4]
```diff
@@ -242,6 +236,12 @@ int aliceVision_main(int argc, char* argv[])
         mesh.generateNormalAndHeightMaps(mp, denseMesh, outputFolder, bumpMappingParams);
     }
 
+    // save final obj file
```
Comment (serguei-k) [inline src/software/pipeline/main_texturing.cpp:239]: moved after texture generation for consistency in logic

**Comment by simogasp** [comment]
Thanks for your contribution!
Would you mind adding the cmake instructions to install the dependency (https://github.com/PixarAnimationStudios/USD I guess) in the main CMakeLists.txt through ExternalProject_Add, as we did for all the others https://github.com/alicevision/AliceVision/blob/develop/CMakeLists.txt#L10?
you can declare a 
```cmake
option(AV_BUILD_USD "Enable building an embedded USD library" ON)
...
message(STATUS "AV_BUILD_USD: ${AV_BUILD_USD}")
```
and 

```cmake
if(AV_BUILD_USD)
set(USD_TARGET usd)
ExternalProject_Add(${USD_TARGET}
       .... # see others for examples
       DEPENDS ${TBB_TARGET} ${BOOST_TARGET}. # these seem to be the only dependencies
       )
set(USD_CMAKE_FLAGS -DUSD_DIR:PATH=${CMAKE_INSTALL_PREFIX}/<where to find the USDConfig.cmake>)
endif()
````
and add `USD_CMAKE_FLAGS` at the very end around line 1059
```cmake
set(AV_DEPS
  ${ZLIB_TARGET}
 .....
  ${CLP_TARGET}
  ${USD_TARGET}       # <<<
)

if(AV_BUILD_ALICEVISION)
ExternalProject_Add(aliceVision
       PREFIX ${CMAKE_CURRENT_SOURCE_DIR}
       .
       .
       .
            ${ZLIB_CMAKE_FLAGS}
            ${ASSIMP_CMAKE_FLAGS}
            ${EIGEN_CMAKE_FLAGS}
            .
            .
            ${FLANN_CMAKE_FLAGS}
            ${PCL_CMAKE_FLAGS}
            ${USD_CMAKE_FLAGS}       # <<<
            -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR> <SOURCE_DIR>
       DEPENDS ${AV_DEPS}
       )
endif()

```

**Comment by fabiencastan** [comment]
_The error in the CI is unrelated to your PR. It's an issue in the CI env since we switched to C++17._

**Comment by serguei-k** [comment]
> Thanks for your contribution!
> 
> Would you mind adding the cmake instructions to install the dependency (https://github.com/PixarAnimationStudios/USD I guess) in the main CMakeLists.txt through ExternalProject_Add, as we did for all the others https://github.com/alicevision/AliceVision/blob/develop/CMakeLists.txt#L10? you can declare a
> 
> ```cmake
> option(AV_BUILD_USD "Enable building an embedded USD library" ON)
> ...
> message(STATUS "AV_BUILD_USD: ${AV_BUILD_USD}")
> ```
> 
> and
> 
> ```cmake
> if(AV_BUILD_USD)
> set(USD_TARGET usd)
> ExternalProject_Add(${USD_TARGET}
>        .... # see others for examples
>        DEPENDS ${TBB_TARGET} ${BOOST_TARGET}. # these seem to be the only dependencies?
>        )
> set(USD_CMAKE_FLAGS -DUSD_DIR:PATH=${CMAKE_INSTALL_PREFIX}/<where to find the USDConfig.cmake>)
> endif()
> ```
> 
> and add `USD_CMAKE_FLAGS` at the very end around line 1059
> 
> ```cmake
> set(AV_DEPS
>   ${ZLIB_TARGET}
>  .....
>   ${CLP_TARGET}
>   ${USD_TARGET}       # <<<
> )
> 
> if(AV_BUILD_ALICEVISION)
> ExternalProject_Add(aliceVision
>        PREFIX ${CMAKE_CURRENT_SOURCE_DIR}
>        .
>        .
>        .
>             ${ZLIB_CMAKE_FLAGS}
>             ${ASSIMP_CMAKE_FLAGS}
>             ${EIGEN_CMAKE_FLAGS}
>             .
>             .
>             ${FLANN_CMAKE_FLAGS}
>             ${PCL_CMAKE_FLAGS}
>             ${USD_CMAKE_FLAGS}       # <<<
>             -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR> <SOURCE_DIR>
>        DEPENDS ${AV_DEPS}
>        )
> endif()
> ```

Thanks for the note, added the optional build step for USD.

**Comment by serguei-k** [comment]
Hi guys, just a heads up that I will be starting a new job next week that will prevent me from working on oss code, so if you would like me to make any changes please let me know :)

**Comment by mugulmd** [comment]
> Hi guys, just a heads up that I will be starting a new job next week that will prevent me from working on oss code, so if you would like me to make any changes please let me know :)

I've just started building your branch for testing, I'll let you know if I spot something 😉

**Comment by simogasp** [comment]
> Hi guys, just a heads up that I will be starting a new job next week that will prevent me from working on oss code, so if you would like me to make any changes please let me know :)

I was wondering whether it could be better to move the part of the code that actually saves the mesh into a USD file (i.e. the code from `main_exportUSD`) directly into the library part, so that it could be used everywhere else if one wants to save a mesh in that format. This would also have the advantage of allowing the meshroom user to save the mesh directly in USD format in the texturing node, without going through another node.

In practice, we could refactor the code and modify the current `Mesh::save()` method so that once it detects the format 
```c++
const std::string fileTypeStr = boost::filesystem::path(filepath).extension().string().substr(1);
const EFileType fileType = mesh::EFileType_stringToEnum(fileTypeStr);
```
and then calls the another save method accordingly: a e.g. `saveUSD()` method that contains the code you already wrote for the `main_exportUSD` or a e.g. `saveAssimp()` that save the mesh using assimp (and that contains the code of the current `Mesh::save()` method).

And, obviously, we can modify all the relevant enum such as `mesh::EFileType` accordingly to add the USD format.

What do you think?

**Review by mugulmd — APPROVED** [review]

**Comment by mugulmd** [comment]
Okay so building AliceVision with USD is not an easy task since there are conflicts with TBB versions, so the fact that this is optional is definitely a plus :sweat_smile: 

I did some testing with textured meshes generated with the photogrammetry pipeline and visualized them in `usdview`, everything seems to be working fine.

Great job, thanks for the contribution!

**Pull request merged**